### DISCUSSION TITLE
Help with replacing the ECDSA cryptography in RETH

### DISCUSSION THREAD

**Discussion by 0xHLiu** [paradigmxyz/reth#9539] [Technical discussions]
Hi, I'm doing a masters thesis on looking at the performance of ethereum if it were to use a post-quantum encryption scheme, so I'm looking to replace the secp256k1 scheme with something else. Going through the entire repo, it looks incredibly deeply ingrained through the entire project. I understand though that the RETH project is very modular, so I think that maybe I'm not looking correctly at which parts of it I should be focusing on. What may be a better way to go about it or a good place to look for what a minimally viable implementation of RETH would look like (mainly which crates I can ignore). Thanks and would appreciate any advice!

**Comment by mattsse** [comment]
cool

> were to use a post-quantum encryption scheme

for what exactly? transactions?

**Reply by 0xHLiu** [reply]
Thanks! Yup, I'm just looking at it for transactions and replacing the digital signatures and keys with FALCON, a lattice cryptography based scheme.

**Reply by mattsse** [reply]
I see, we currently don't have an easy way to configure tx types yet, this is wip, so I'd recommend hacking some additional txtypes in there that use a different signature

**Reply by 0xHLiu** [reply]
Thanks for the fast replies!

What may be the best way to look into the crates to go about it (or mainly finding out which ones I can pretty much ignore completely? I think that's the part I'm having a hard time with. I see from the project layout (https://github.com/paradigmxyz/reth/blob/main/docs/repo/layout.md#primitives) that is it mainly the REVM crate to focus on since it deals with transaction execution, but from doing a quick search, the secp256k1 crate is mentioned pretty much all throughout the whole repository (and not from the REVM crate).

**Reply by onbjerg** [reply]
i'd focus my efforts on the primitives crate first as that is where we define the transaction types incl. signature recovery. the rest of the usage is mostly networking, and in revm afaik the only usage is for precompiles defined in ethereum

**Reply by 0xHLiu** [reply]
Thank you for this tip!

**Reply by 0xHLiu** [reply]
By the way, looking more into this, is public key recovery from the signature and message essential in Reth? I am trying to remove the "recover_signer_unchecked" function from util.rs in the primitives crate, but it looks to be used everywhere across the project. The cryptography scheme I'm trying to replace ECDSA with can't recover the public key from the signature and message.

**Reply by onbjerg** [reply]
It's essential to Ethereum as transactions do not store the sending address. Transactions are just signed, and using the signature hash of the transaction and the signature, we recover the public key and from that the address

**Reply by 0xHLiu** [reply]
Thanks for this! 
Is there an easy way to find how which functions in RETH are called when using RETH in a private net? For example, I've followed the kurtosis tutorial to set up a private network only uses RETH and lighthouse for the consensus layer. If I'm on foundry, and call the function to transfer ETH from one address to another, how do I see what functions are called within the RETH package?
The cryptography scheme I'm looking at (FALCON) doesn't have public key recovery so I have to remove a lot of the functions associated with the Signature Struct, and I want to see if there is a simple way possible to see if the network will still work without a lot of those functions. If it's hard to check, I can one-by-one deleted some functions and set it up on kurtosis, but it takes about 2hours at this point to dockerize RETH to get it set up on kurtosis.

**Reply by onbjerg** [reply]
Hi, there is no full stack trace like this, unless you use something like `perf` to record, but it would be incomplete and very noisy.

If there is no way to recover the public key, it's going to take a lot of changes to make it work with Ethereum. A transaction is encoded, then hashed, and then signed. All Reth gets is the transaction body with the signature (r, s, v), and we need to recover the public key in order to determine what account to charge for the transaction upon inclusion. You would need to figure out some alternative scheme for this, and also adjust the `to` field of the transactions in order to identify the recipient of a transaction, too (e.g. for contract execution or for balance transfers). Since the address scheme would need to be adjusted, this has ripple effects on e.g. the block header, where there is an address to determine who gets rewards for block building and inclusion.

As another note, Reth keys account data by the account address in the database.

Lastly, I think this probably also affects the CL in terms of deposits/withdrawals.
